File: Leetcode/Swift/0014. Longest Common Prefix.swift
//
//  Solution.swift
//  LeetCodeScrachSw
//
//  Created by Richard Cho on 2022/09/26.
//

import Foundation

class Solution {
    func longestCommonPrefix(_ strs: [String]) -> String {
        if strs.isEmpty { return "" }
        if strs.count == 1 { return strs[0] }

        var shortest = "", prefix = ""
        var shortestLength = Int.max
        
        for str in strs {
            if str.count < shortestLength {
                shortestLength = str.count
                shortest = str
            }
        }
        
        for i in 0...shortestLength {
            prefix = shortest[0, i]
            
            for j in 0..<strs.count {
                if !strs[j].starts(with: prefix) {
                    return prefix[0, prefix.count - 1]
                }
            }
        }
                
        return prefix;
    }
}


extension String {
    
    subscript(start: Int, end: Int) -> String {
        get {
            let startIndex = self.index(self.startIndex, offsetBy: start)
            let endIndex = self.index(self.startIndex, offsetBy: end)
            return String(self[startIndex..<endIndex])
        }
        set {
            let startIndex = self.index(self.startIndex, offsetBy: start)
            let endIndex = self.index(self.startIndex, offsetBy: end)
            self = self.replacingCharacters(in: startIndex..<endIndex, with: newValue)
        }
    }
    
}

File: Leetcode/Java/0014. Longest Common Prefix.java
public class Solution {
    public String longestCommonPrefix(String[] strs) {
        if (strs.length == 0) { return ""; }
        if (strs.length == 1) { return strs[0]; }

        String shortest = "", prefix = "";
        int shortestLength = Integer.MAX_VALUE;

        for (String str: strs) {
            int len = str.length();
            if (len < shortestLength) {
                shortestLength = str.length();
                shortest = str;
            }
        }

        for (int i = 0; i <= shortestLength; i++) {
            prefix = shortest.substring(0, i);
            for (int j = 0; j < strs.length; j++) {
                if (!strs[j].startsWith(prefix)) {
                    return prefix.substring(0, prefix.length() - 1);
                }
            }
        }

        return prefix;
    }
}